### PR TITLE
Display view licence tabs only to permitted users

### PR DESCRIPTION
**Pull request by jonathangoulding**
Display view licence tabs only to permitted users

https://eaflood.atlassian.net/browse/WATER-4432

> Part of a series of changes to replace the legacy view licence page

This change will only show the billing and charge information tabs on the view license page if the user has the appropriate role.

### CODE REVIEW

File: app/controllers/licences.controller.js [hunk 1]
```diff
@@ -25,9 +25,9 @@ async function returnsRequired (request, h) {
 }
 
 async function view (request, h) {
-  const { id } = request.params
+  const { params: { id }, auth } = request
 
-  const data = await ViewLicenceService.go(id)
+  const data = await ViewLicenceService.go(id, auth)
```
Comment (jonathangoulding) [inline app/controllers/licences.controller.js:30]: I could not see the benefit adding to this test as the ViewLicenceService is mocked. 

I can not see other places where the the mocks args are tested in the rest of the code base. 

I think the presenter test is sufficient to test the auth functionality but happy to discuss.

File: app/presenters/licences/view-licence.presenter.js [hunk 2]
```diff
@@ -71,7 +71,8 @@ function go (licence, licenceAbstractionConditions) {
     registeredTo,
     sourceOfSupply: abstractionDetails.sourceOfSupply,
     startDate: formatLongDate(startDate),
-    warning: _generateWarningMessage(ends)
+    warning: _generateWarningMessage(ends),
+    roles: _getAuthRoles(auth)
```
Comment (jonathangoulding) [inline app/presenters/licences/view-licence.presenter.js:75]: Started to get a sonar cloud error - https://sonarcloud.io/project/issues?resolved=false&severities=MAJOR&sinceLeakPeriod=true&branch=view-license-role-based-tabs-for-bills-and-charge-information&id=DEFRA_water-abstraction-system&open=AY8pIrHhp7fcpGn2GYJe

this did not happen when the notification work was done.

File: app/controllers/licences.controller.js (hunk 1 above)
Comment (Cruikshanks) [reply app/controllers/licences.controller.js:30]: Agreed. This is why our controllers are 'dumb' and leave all the work to the services. In general, we limit our controller tests to 'when the request is valid it [returns x/redirects to y]' and 'when request is invalid it [returns z]'

Any other tests of behaviour are done directly against the service/presenter.

File: test/presenters/licences/view-licence.presenter.test.js [hunk 3]
```diff
@@ -1120,6 +1123,40 @@ describe('View Licence presenter', () => {
...
+
+        expect(result.roles).to.equal(['role 1', 'role 2'])
+      })
+    })
+    describe('when the authenticated user has NO roles', () => {
+      beforeEach(() => {
+        auth = undefined
+      })
+
+      it('returns the roles in a flat array', () => {
```
Comment (Cruikshanks) [inline test/presenters/licences/view-licence.presenter.test.js:1153]: Guessing a copy & paste typo

```suggestion
      it('returns null for the roles', () => {
```

File: app/presenters/licences/view-licence.presenter.js [hunk 4]
```diff
@@ -132,6 +133,12 @@ function _endDate (expiredDate) {
   return formatLongDate(expiredDate)
 }
 
+function _getAuthRoles (auth) {
+  return auth?.credentials?.roles?.map((role) => {
+    return role?.role
+  }) || null
```
Comment (Cruikshanks) [inline app/presenters/licences/view-licence.presenter.js:139]: Having experienced being thrown at a Node.js project when I knew very little about the language I try and keep things relatively simple.

In this case, I'd just roll it back one step and suggest this

```suggestion
  const roles = auth?.credentials?.roles?.map((role) => {
    return role?.role
  })

  return roles || null
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
Just a few corrections and suggestions!

File: app/controllers/licences.controller.js [hunk 5]
```diff
@@ -25,9 +25,9 @@ async function returnsRequired (request, h) {
 }
 
 async function view (request, h) {
-  const { id } = request.params
+  const { params: { id }, auth } = request
```
[the remark this answers is not in the document]
Comment (jonathangoulding) [reply app/controllers/licences.controller.js:28]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#nested_object_and_array_destructuring

For one nested object its very useful. 

If this was  `const { params: { id, fella }, auth } = request` and `fella` did not exist then fella would be undefined. 

If this was  `const { params: { id, fella: { thing }  }, auth } = request` and `thing` did not exist then if would error: 
"Cannot read properties of undefined (reading 'thing')"

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**